connector/spanmetricsconnector/connector.go

```diff
@@ -0,0 +1,544 @@
+	logger.Info("Building spanmetrics")
+	pConfig := config.(*Config)
+
+	bounds := defaultLatencyHistogramBucketsMs
+	if pConfig.LatencyHistogramBuckets != nil {
+		bounds = mapDurationsToMillis(pConfig.LatencyHistogramBuckets)
+	}
+
+	if err := validateDimensions(pConfig.Dimensions, pConfig.skipSanitizeLabel); err != nil {
+		return nil, err
+	}
+
+	if pConfig.DimensionsCacheSize <= 0 {
+		return nil, fmt.Errorf(
+			"invalid cache size: %v, the maximum number of the items in the cache should be positive",
+			pConfig.DimensionsCacheSize,
+		)
+	}
+	metricKeyToDimensionsCache, err := cache.NewCache[metricKey, pcommon.Map](pConfig.DimensionsCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	return &connectorImp{
+		logger:                logger,
+		config:                *pConfig,
+		startTimestamp:        pcommon.NewTimestampFromTime(time.Now()),
+		latencyBounds:         bounds,
+		histograms:            make(map[metricKey]*histogramData),
+		dimensions:            newDimensions(pConfig.Dimensions),
+		keyBuf:                bytes.NewBuffer(make([]byte, 0, 1024)),
+		metricKeyToDimensions: metricKeyToDimensionsCache,
+		ticker:                ticker,
+		done:                  make(chan struct{}),
+	}, nil
+}
+
+// durationToMillis converts the given duration to the number of milliseconds it represents.
+// Note that this can return sub-millisecond (i.e. < 1ms) values as well.
+func durationToMillis(d time.Duration) float64 {
+	return float64(d.Nanoseconds()) / float64(time.Millisecond.Nanoseconds())
+}
+
+func mapDurationsToMillis(vs []time.Duration) []float64 {
+	vsm := make([]float64, len(vs))
+	for i, v := range vs {
+		vsm[i] = durationToMillis(v)
+	}
+	return vsm
+}
+
+// validateDimensions checks duplicates for reserved dimensions and additional dimensions. Considering
+// the usage of Prometheus related exporters, we also validate the dimensions after sanitization.
+func validateDimensions(dimensions []Dimension, skipSanitizeLabel bool) error {
+	labelNames := make(map[string]struct{})
+	for _, key := range []string{serviceNameKey, spanKindKey, statusCodeKey} {
+		labelNames[key] = struct{}{}
+		labelNames[sanitize(key, skipSanitizeLabel)] = struct{}{}
+	}
+	labelNames[operationKey] = struct{}{}
+
+	for _, key := range dimensions {
+		if _, ok := labelNames[key.Name]; ok {
+			return fmt.Errorf("duplicate dimension name %s", key.Name)
+		}
+		labelNames[key.Name] = struct{}{}
+
+		sanitizedName := sanitize(key.Name, skipSanitizeLabel)
+		if sanitizedName == key.Name {
+			continue
+		}
+		if _, ok := labelNames[sanitizedName]; ok {
+			return fmt.Errorf("duplicate dimension name %s after sanitization", sanitizedName)
+		}
+		labelNames[sanitizedName] = struct{}{}
+	}
+
+	return nil
+}
+
+// Start implements the component.Component interface.
+func (p *connectorImp) Start(ctx context.Context, host component.Host) error {
+	p.logger.Info("Starting spanmetricsprocessor")
```

Comment:
```suggestion
	p.logger.Info("Starting spanmetric connector")
```

Please update the rest of log entries as well.